components/client/src/main/java/com/hotels/styx/client/loadbalancing/strategies/RoundRobinStrategy.java

```diff
@@ -46,56 +46,36 @@
+public class RoundRobinStrategy implements LoadBalancer {
 
-    private final ActiveOrigins activeOrigins;
+    private ActiveOrigins activeOrigins;
+    private final AtomicReference<ArrayList<RemoteHost>> origins;
+    private final AtomicInteger index = new AtomicInteger(0);
 
-    public RoundRobinStrategy(ActiveOrigins activeOrigins) {
-        this.activeOrigins = activeOrigins;
+    public RoundRobinStrategy(ActiveOrigins activeOrigins, Iterable<RemoteHost> initialOrigins) {
+        this.activeOrigins = requireNonNull(activeOrigins);
+        origins = new AtomicReference<>(new ArrayList<>(newArrayList(initialOrigins)));
     }
 
     /**
      * Factory for creating {@link com.hotels.styx.client.loadbalancing.strategies.RoundRobinStrategy}.
      */
-    public static class Factory implements LoadBalancingStrategyFactory {
+    public static class Factory implements LoadBalancerFactory {
         @Override
-        public LoadBalancingStrategy create(Environment environment, Configuration strategyConfiguration, ActiveOrigins activeOrigins) {
-            return new RoundRobinStrategy(activeOrigins);
+        public LoadBalancer create(Environment environment, Configuration strategyConfiguration, ActiveOrigins activeOrigins) {
+            return new RoundRobinStrategy(activeOrigins, activeOrigins.snapshot());
         }
     }
 
-    private final AtomicInteger index = new AtomicInteger(0);
-
     @Override
-    public Iterable<RemoteHost> vote(Context context) {
-        Iterable<RemoteHost> snapshot = activeOrigins.snapshot();
-        return isEmpty(snapshot) ? snapshot : cycledNonExhaustedOrigins(snapshot);
-    }
-
-    private List<RemoteHost> cycledNonExhaustedOrigins(Iterable<RemoteHost> origins) {
-        return cycleOrigins(origins)
-                .filter(host -> !host.connectionPool().isExhausted())
-                .collect(toList());
+    public Optional<RemoteHost> choose(Preferences preferences) {
+        ArrayList<RemoteHost> remoteHosts = origins.get();
+        return Optional.ofNullable(remoteHosts.get(index.getAndIncrement() % remoteHosts.size()));
     }
 
     @Override
-    public void originsInventoryStateChanged(OriginsInventorySnapshot snapshot) {
-        index.set(0);
-    }
-
-    private Stream<RemoteHost> cycleOrigins(Iterable<RemoteHost> origins) {
-        List<RemoteHost> originsList = newArrayList(origins);
-        return cycleToOffset(nextIndex(originsList), originsList);
-    }
-
-    private int nextIndex(List<RemoteHost> origins) {
-        return index.getAndIncrement() % origins.size();
-    }
-
-    private static Stream<RemoteHost> cycleToOffset(int index, List<RemoteHost> origins) {
-        List<RemoteHost> first = origins.subList(index, origins.size());
-        List<RemoteHost> second = origins.subList(0, index);
-        return concat(first.stream(), second.stream());
+    public void originsChanged(OriginsSnapshot snapshot) {
```

Comment:
The event `originsChanged` triggers origins update via a call to `activeOrigins.snapshot()`. The `snapshot` comes from the interface. Therefore it must be present. The `snapshot` is not used because it delivers the hosts as `Set<Origin>` but what we really needs is a `Set<RemoteHost>`.

In near future we need to change the snapshot origins list signature to `Set<Origin>` instead.